File: app/src/main/java/com/example/rahmatantravel/Adapter/PaketAdapter.java
package com.example.rahmatantravel.Adapter;

import android.text.Layout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.rahmatantravel.Models.PaketHomeModels;
import com.example.rahmatantravel.Models.PaketModels;
import com.example.rahmatantravel.R;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;

public class PaketAdapter extends RecyclerView.Adapter<PaketAdapter.PaketViewHolder> {
    private ArrayList<PaketModels> dataList;
//    private ArrayList<PaketModels> filteredList;

    public PaketAdapter(ArrayList<PaketModels> dataList)
    {
        this.dataList = dataList;
//        this.filteredList = new ArrayList<>(dataList);
    }

//    public void setFilteredList(List<PaketModels> filteredList){
//        this.filteredList = (ArrayList<PaketModels>) filteredList;
//        notifyDataSetChanged();
//    }

    @NonNull
    @Override
    public PaketAdapter.PaketViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType){
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        View view = layoutInflater.inflate(R.layout.item_paket, parent, false);
        return new PaketViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull PaketAdapter.PaketViewHolder holder, int position) {
        holder.judulPaket.setText(dataList.get(position).getJudulPaket());

        SimpleDateFormat dateFormat = new SimpleDateFormat("dd MMMM yyyy");
        String formattedDate = dateFormat.format(dataList.get(position).getTanggalBerangkat());
        holder.tanggalBerangkat.setText(formattedDate);
        holder.waktuPaket.setText(String.valueOf(dataList.get(position).getWaktuPaket()));
        holder.rateHotel.setText(String.valueOf(dataList.get(position).getRateHotel()));
        holder.hargaPaket.setText(String.valueOf(dataList.get(position).getHargaPaket()));
    }

    @Override
    public int getItemCount() {
        return (dataList != null) ? dataList.size() : 0;
    }

    public class PaketViewHolder extends RecyclerView.ViewHolder {
        private TextView judulPaket, tanggalBerangkat, waktuPaket, rateHotel,hargaPaket;

        public PaketViewHolder(View view) {
            super(view);
            judulPaket = (TextView) itemView.findViewById(R.id.judulPaket);
            tanggalBerangkat = (TextView) itemView.findViewById(R.id.tanggalBerangkat);
            waktuPaket = (TextView) itemView.findViewById(R.id.waktuPaket);
            rateHotel = (TextView) itemView.findViewById(R.id.rateHotel);
            hargaPaket = (TextView) itemView.findViewById(R.id.hargaPaket);
        }
    }
}

File: app/src/main/java/com/example/rahmatantravel/Adapter/PaketAdapter.kt
package com.example.rahmatantravel.Adapter;

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.rahmatantravel.Models.PaketModels
import com.example.rahmatantravel.R
import com.example.rahmatantravel.api.paketResponse.APIResponse
import com.example.rahmatantravel.api.paketResponse.HotelResponse
import com.example.rahmatantravel.api.paketResponse.KeberangkatanResponse
import com.example.rahmatantravel.api.paketResponse.PaketResponse
import java.text.NumberFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class PaketAdapter(private var keberangkatanResponse: List<KeberangkatanResponse>,
                   private var paketResponse: List<PaketResponse>,
                   private var hotelResponse: List<HotelResponse>) : RecyclerView.Adapter<PaketAdapter.PaketViewHolder>() {

    var onItemClick : ((KeberangkatanResponse, PaketResponse, HotelResponse) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PaketViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val view = layoutInflater.inflate(R.layout.item_paket, parent, false)
        return PaketViewHolder(view)
    }

    override fun onBindViewHolder(holder: PaketViewHolder, position: Int) {
        val keberangkatanItem = keberangkatanResponse[position]
        val paketItem = paketResponse[position]
        val hotelItem = hotelResponse[position]
        holder.bind(keberangkatanItem, paketItem, hotelItem)
        holder.itemView.setOnClickListener{
            onItemClick?.invoke(keberangkatanItem, paketItem, hotelItem)
        }
    }

    override fun getItemCount(): Int {
        return minOf(keberangkatanResponse.size, paketResponse.size, hotelResponse.size)
    }

    fun setFilteredList(keberangkatanResponse: List<KeberangkatanResponse>, paketResponse: List<PaketResponse>, hotelResponse: List<HotelResponse>) {
        this.keberangkatanResponse = keberangkatanResponse
        this.paketResponse = paketResponse
        this.hotelResponse = hotelResponse
        notifyDataSetChanged()
    }

    fun getKeberangkatanResponse(): List<KeberangkatanResponse> {
        return keberangkatanResponse
    }

    fun getPaketResponse(): List<PaketResponse> {
        return paketResponse
    }

    fun getHotelResponse(): List<HotelResponse> {
        return hotelResponse
    }

    inner class PaketViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val judulPaket: TextView = itemView.findViewById(R.id.judulPaket)
        val tanggalBerangkat: TextView = itemView.findViewById(R.id.tanggalBerangkat)
        val waktuPaket: TextView = itemView.findViewById(R.id.waktuPaket)
        val rateHotel: TextView = itemView.findViewById(R.id.rateHotel)
        val hargaPaket: TextView = itemView.findViewById(R.id.hargaPaket)
        val gambarPaket: ImageView = itemView.findViewById(R.id.brosurPaket)

        fun bind(keberangkatanResponse: KeberangkatanResponse, paketResponse: PaketResponse, hotelResponse: HotelResponse){
            val tanggalPaketFormatted = formatDate(keberangkatanResponse.tanggal)

            val harga = formatCurrency(paketResponse.harga[0].harga)

            judulPaket.text = paketResponse.nama_paket
            tanggalBerangkat.text = tanggalPaketFormatted
            waktuPaket.text = paketResponse.lama_hari.toString()
            rateHotel.text = hotelResponse.bintang.toString()
            hargaPaket.text = "Mulai dari $harga"

            Glide.with(itemView.context)
                .load("https://rahmatanumrah.000webhostapp.com/uploads/foto_brosur/${paketResponse.foto_paket}")
                .into(gambarPaket)
        }

        fun formatDate(inputDate: String): String {
            val inputFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
            val outputFormat = SimpleDateFormat("dd MMMM yyyy", Locale.getDefault())
            val date: Date = inputFormat.parse(inputDate) ?: Date()
            return outputFormat.format(date)
        }

        fun formatCurrency(number: Int): String {
            val format = NumberFormat.getCurrencyInstance(Locale("id", "ID")) // Locale untuk Indonesia
            return format.format(number)
        }
    }
}

